packages/Popover/src/Popover.js

```diff
@@ -0,0 +1,517 @@
+
+    if (top && left && transformOrigin && transform) {
+      return {
+        top,
+        left,
+        transformOrigin,
+        transform,
+      };
+    } 
+  }
+
+  // Determines the alignment to render based on an order of alignment fallbacks
+  // Returns the first alignment that doesn't collide with the window,
+  // defaulting to the align prop if all alignments fail.
+  getAlignment() {
+    const { align } = this.props;
+    const alignments = {
+      top: ['top', 'bottom'],
+      bottom: ['bottom', 'top'],
+      left: ['left', 'right'],
+      right: ['right', 'left'],
+    };
+
+    let validAlignment = align;
+
+    alignments[align].some(candidate => {
+      if (!this.checkAlignment(candidate)) {
+        return false;
+      }
+
+      validAlignment = candidate;
+      return true;
+    });
+
+    return validAlignment;
+  }
+
+  // Determines the justification to render based on an order of justification fallbacks
+  // Returns the first justification that doesn't collide with the window,
+  // defaulting to the justify prop if all justifications fail.
+  getJustification(alignment) {
+    const { justify } = this.props;
+    let justifications = {};
+
+    switch (alignment) {
+      case 'top':
+      case 'bottom': {
+        justifications = {
+          start: ['left', 'right', 'center'],
+          middle: ['center-horizontal', 'right', 'left'],
+          end: ['right', 'left', 'center'],
+        };
+        break;
+      }
+
+      case 'left':
+      case 'right': {
+        justifications = {
+          start: ['top', 'bottom', 'center'],
+          middle: ['center-vertical', 'bottom', 'top'],
+          end: ['bottom', 'top', 'center'],
+        };
+        break;
+      }
+    }
+
+    let validJustification = false;
+
+    justifications[justify].some(candidate => {
+      if (!this.checkJustification(candidate)) {
+        return false;
+      }
+
+      validJustification = candidate;
+      return true;
+    });
+
+    return validJustification;
+  }
+
+  // Checks that an alignment will not cause the popover to collide with the window.
+  checkAlignment(alignment) {
+    const { withoutPortal } = this.props;
+    const top = withoutPortal
+      ? this.calcWithoutPortalTop({ alignment })
+      : this.calcTop({ alignment });
+    const left = withoutPortal
+      ? this.calcWithoutPortalLeft({ alignment })
+      : this.calcLeft({ alignment });
+
+    if (['top', 'bottom'].includes(alignment)) {
+      return this.checkVerticalWindowCollision(top);
+    }
+
+    if (['left', 'right'].includes(alignment)) {
+      return this.checkHorizontalWindowCollision(left);
+    }
+  }
+
+  // Checks that a justification will not cause the popover to collide with the window.
+  checkJustification(justification) {
+    const { withoutPortal } = this.props;
+    const top = withoutPortal
+      ? this.calcWithoutPortalTop({ justification })
+      : this.calcTop({ justification });
+    const left = withoutPortal
```

Comment:
All this ternary block-style formatting is hard to read without spaces in between. Could we be sure to add trailing spaces after blocks, even these?